cmd/pullrequest-init/disk_test.go

```diff
@@ -0,0 +1,289 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestToDisk(t *testing.T) {
+	tektonPr := PullRequest{
+		Type: "github",
+		ID:   123,
+		Head: &GitReference{
+			Repo:   "foo1",
+			Branch: "branch1",
+			SHA:    "sha1",
+		},
+		Base: &GitReference{
+			Repo:   "foo2",
+			Branch: "branch2",
+			SHA:    "sha2",
+		},
+		Statuses: []*Status{
+			{
+				ID:          "123",
+				Code:        Success,
+				Description: "foobar",
+				URL:         "https://foo.bar",
+			},
+		},
+		Comments: []*Comment{
+			{
+				Text:   "hey",
+				Author: "me",
+				ID:     123,
+			},
+		},
+		Labels: []*Label{
+			{Text: "help"},
+			{Text: "me"},
+		},
+	}
+
+	d, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(d)
+	fmt.Println(d)
+	if err := ToDisk(&tektonPr, d); err != nil {
+		t.Error(err)
+	}
+
+	// Check the refs
+	checkRef := func(name string, r GitReference) {
+		actualRef := GitReference{}
+		readAndUnmarshal(t, filepath.Join(d, name), &actualRef)
+		if diff := cmp.Diff(actualRef, r); diff != "" {
+			t.Errorf("Get PullRequest: -want +got: %s", diff)
+		}
+	}
+	checkRef("head", *tektonPr.Head)
+	checkRef("base", *tektonPr.Base)
+
+	// Check the Statuses
+	fis, err := ioutil.ReadDir(filepath.Join(d, "status"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	statuses := map[string]Status{}
+	for _, fi := range fis {
+		status := Status{}
+		readAndUnmarshal(t, filepath.Join(d, "status", fi.Name()), &status)
+		statuses[status.ID] = status
+	}
+	for _, s := range tektonPr.Statuses {
+		actualStatus, ok := statuses[s.ID]
+		if !ok {
+			t.Errorf("Expected status with ID: %s, not found: %v", s.ID, statuses)
+		}
+		if diff := cmp.Diff(actualStatus, *s); diff != "" {
+			t.Errorf("Get Status: -want +got: %s", diff)
+		}
+	}
+
+	// Check the labels
+	fis, err = ioutil.ReadDir(filepath.Join(d, "labels"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	labels := map[string]struct{}{}
+	for _, fi := range fis {
+		labels[fi.Name()] = struct{}{}
+	}
+
+	for _, l := range tektonPr.Labels {
+		if _, ok := labels[l.Text]; !ok {
+			t.Errorf("Expected label with text: %s, not found: %v", l.Text, labels)
+		}
+	}
+
+	// Check the comments
+	fis, err = ioutil.ReadDir(filepath.Join(d, "comments"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	comments := map[int64]Comment{}
+	for _, fi := range fis {
+		comment := Comment{}
+		readAndUnmarshal(t, filepath.Join(d, "comments", fi.Name()), &comment)
+		comments[comment.ID] = comment
+	}
+	for _, c := range tektonPr.Comments {
+		actualComment, ok := comments[c.ID]
+		if !ok {
+			t.Errorf("Expected comment with ID: %d, not found: %v", c.ID, comments)
+		}
+		if diff := cmp.Diff(actualComment, *c); diff != "" {
+			t.Errorf("Get Comment: -want +got: %s", diff)
+		}
+	}
+}
```

Comment:
the way this test and this code are broken up makes me think we could break up `ToDisk` even more into some separate methods with some separate tests - pretty much every section that's like "now we check x" "now we check y"

Reply:
Done. I don't actually think I like separating out the tests though.. The logic of each one is very simple, and if we tested at the focused level we'd still want to test at the overall level to make sure the overall function didn't forget to call one of the sub functions.

LMK what you think...

Reply:
i agree with still having one test at the top level  - however if it was up to me id still make individual tests, then make the test of the top level function as simple as possible (maybe even just going to disk and back again, making sure some of the expected values were present). the approach im suggesting probably does have more potential to miss something, but i think in conjunction with an end to end test that actually uses this, we'd have pretty much everything covered.

i find individual tests much easier to read and maintain over time, and easier to make sure that everything is covered as desired (e.g. reading a small function + writing a test for it vs. reading a bunch of small functions while reading through one big test)

but i think we're getting very close to a situation where its just that i would have done it differently from you so im happy with it the way it is XD

if we need to start changing those individual functions them maybe we can revisit it

Reply:
oh here's a thought tho: how do you plan to add in gitlab support? there could be some argument made for logic like so:

You could have `FromDisk` and `ToDisk` functions that do all of:
- comments
- labels
- statuses
- refs

But operate on different implementations of comments, labels, statuses and refs for each of GitHub and GitLab. In that case you might want some top level test that makes sure that FromDisk and ToDisk are operating correctly, but then separate unit tests for each of comments, labels, statuses, refs for each implementation?

I dunno just thoughts